shell/platform/glfw/glfw_event_loop.cc

```diff
@@ -0,0 +1,114 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/shell/platform/glfw/glfw_event_loop.h"
+
+#include <GLFW/glfw3.h>
+
+#include <atomic>
+#include <utility>
+
+namespace flutter {
+
+GLFWEventLoop::GLFWEventLoop(std::thread::id main_thread_id,
+                             TaskExpiredCallback on_task_expired)
+    : main_thread_id_(main_thread_id),
+      on_task_expired_(std::move(on_task_expired)) {}
+
+GLFWEventLoop::~GLFWEventLoop() = default;
+
+bool GLFWEventLoop::RunsTasksOnCurrentThread() const {
+  return std::this_thread::get_id() == main_thread_id_;
+}
+
+void GLFWEventLoop::WaitForEvents(std::chrono::nanoseconds max_wait) {
+  const auto now = TaskTimePoint::clock::now();
+  std::vector<FlutterTask> expired_tasks;
+
+  // Process exipred tasks.
+  {
+    std::lock_guard<std::mutex> lock(task_queue_mutex_);
+    while (!task_queue_.empty()) {
+      const auto& top = task_queue_.top();
+      // If this task (and all tasks after this) has not yet expired, there is
+      // nothing more to do. Quit iterating.
+      if (top.fire_time > now) {
+        break;
+      }
+
+      // Make a record of the expired task. Do NOT service the task here
+      // because we are still holding onto the task queue mutex. We don't want
+      // other threads to block on posting tasks onto this thread till we are
+      // done processing expired tasks.
+      expired_tasks.push_back(task_queue_.top().task);
+
+      // Remove the tasks from the delayed tasks queue.
+      task_queue_.pop();
+    }
+  }
+
+  // Fire expired tasks.
+  {
```

Comment:
There's no scoped lock; why is there a scope block? (I would argue that the fact that you did this despite in not being any mechanical reason I can see is an indication that it should be broken into helpers.)